File: LeetcodeCpp/FindAllPeopleWithSecret.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    #define ppr pair<int, int>
    vector<int> findAllPeople(int n, vector<vector<int>>& meetings, int firstPerson) {
        vector<vector<pair<int,int>>> adj(n);
        for (auto it : meetings) {
            adj[it[0]].push_back({it[1],it[2]});
            adj[it[1]].push_back({it[0],it[2]});
        }
        priority_queue<ppr, vector<ppr>, greater<ppr> > pq;
        pq.push({0, 0});
        pq.push({0, firstPerson});

        vector<int> vis(n, 0);

        while (!pq.empty()) {
            auto it = pq.top();
            int time=it.first; 
            int person=it.second;
            pq.pop();

            if (vis[person]) {
                continue;
            }
            vis[person] = true;
            for (auto it : adj[person]) {
                if (!vis[it.first] && it.second >= time) {
                    pq.push({it.second, it.first});
                }
            }
        }
        vector<int> ans;
        for (int i = 0; i < n; ++i) {
            if (vis[i]) {
                ans.push_back(i);
            }
        }
        return ans;
    }
};

int main() {
    Solution solution;

    // Test case 1
    int n1 = 4;
    vector<vector<int>> meetings1 = {{0, 1, 5}, {0, 2, 7}, {1, 3, 6}, {2, 3, 8}};
    int firstPerson1 = 0;
    vector<int> result1 = solution.findAllPeople(n1, meetings1, firstPerson1);
    cout << "Test Case 1 Result: ";
    for (int person : result1) {
        cout << person << " ";
    }
    cout << endl;

    // Test case 2
    int n2 = 5;
    vector<vector<int>> meetings2 = {{0, 1, 5}, {0, 2, 7}, {1, 3, 6}, {2, 3, 8}, {3, 4, 9}};
    int firstPerson2 = 1;
    vector<int> result2 = solution.findAllPeople(n2, meetings2, firstPerson2);
    cout << "Test Case 2 Result: ";
    for (int person : result2) {
        cout << person << " ";
    }
    cout << endl;

    return 0;
}

File: LeetcodeJava/FindAllPeopleWithSecret.java
import java.util.*;

class Solution {
    class Pair {
        int first;
        int second;
        
        Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }
    }
    
    public List<Integer> findAllPeople(int n, int[][] meetings, int firstPerson) {
        List<List<Pair>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        
        for (int[] it : meetings) {
            adj.get(it[0]).add(new Pair(it[1], it[2]));
            adj.get(it[1]).add(new Pair(it[0], it[2]));
        }
        
        PriorityQueue<Pair> pq = new PriorityQueue<>(Comparator.comparingInt(p -> p.first));
        pq.add(new Pair(0, 0));
        pq.add(new Pair(0, firstPerson));

        boolean[] vis = new boolean[n];

        while (!pq.isEmpty()) {
            Pair it = pq.poll();
            int time = it.first;
            int person = it.second;

            if (vis[person]) {
                continue;
            }
            vis[person] = true;
            for (Pair pair : adj.get(person)) {
                if (!vis[pair.first] && pair.second >= time) {
                    pq.add(new Pair(pair.second, pair.first));
                }
            }
        }
        List<Integer> ans = new ArrayList<>();
        for (int i = 0; i < n; ++i) {
            if (vis[i]) {
                ans.add(i);
            }
        }
        return ans;
    }
    
    public static void main(String[] args) {
        Solution solution = new Solution();

        // Test case 1
        int n1 = 4;
        int[][] meetings1 = {{0, 1, 5}, {0, 2, 7}, {1, 3, 6}, {2, 3, 8}};
        int firstPerson1 = 0;
        List<Integer> result1 = solution.findAllPeople(n1, meetings1, firstPerson1);
        System.out.print("Test Case 1 Result: ");
        for (int person : result1) {
            System.out.print(person + " ");
        }
        System.out.println();

        // Test case 2
        int n2 = 5;
        int[][] meetings2 = {{0, 1, 5}, {0, 2, 7}, {1, 3, 6}, {2, 3, 8}, {3, 4, 9}};
        int firstPerson2 = 1;
        List<Integer> result2 = solution.findAllPeople(n2, meetings2, firstPerson2);
        System.out.print("Test Case 2 Result: ");
        for (int person : result2) {
            System.out.print(person + " ");
        }
        System.out.println();
    }
}
